File: PRACTICA4/ruby/NumericDamage.rb
require_relative 'NumericDamageToUI'
require_relative "Damage"
module Deepspace
  class NumericDamage < Damage
    def initialize(nweapons,nshields)
      super(nshields)
      @nWeapons=nweapons
    end

    def copy
      return self.class.new(@nWeapons,@nShields)
    end

    def nWeapons
      return @nWeapons
    end

    def adjust(w,s)
      return self.class.new([@nWeapons,w.length].min,super(s))
    end

    def discardWeapon(w)
      if(@nWeapons>0)
        @nWeapons-=1
      end
    end

    def hasNoEffect
      return (@nWeapons + @nShields) == 0
    end

    def getUIversion
      return NumericDamageToUI.new(self)
    end
    def to_s
      getUIversion().to_s
    end
  end

end

File: Examen SpaceFighter/src/deepspace/NumericDamage.java
package deepspace;

import java.util.ArrayList;

public class NumericDamage extends Damage{
    int nWeapons;

    public NumericDamage(int nweapons, int nshields){
        super(nshields);
        nWeapons=nweapons;
    }

    @Override
    public NumericDamage copy(){
        return new NumericDamage(nWeapons,getNShields());
    }

    public int getNWeapons(){
        return nWeapons;
    }

    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        int nshields=Integer.min(getNShields(), s.size());
        int nweapons=Integer.min(nWeapons, w.size());

        return new NumericDamage(nweapons,nshields);

    }

    @Override
    public void discardWeapon(Weapon w){
        if(nWeapons>0) nWeapons--;
    }

    @Override
    public boolean hasNoEffect(){
        return (nWeapons + getNShields() == 0);
    }

    @Override
    public String toString(){
        String salida = "[Damage] -> nWeapons: " + nWeapons +
                ", nShields: " + getNShields();

        return salida;
    }


    @Override
    public NumericDamageToUI getUIversion(){
        return new NumericDamageToUI(this);
    }


}